File: PyLibs/src/Libs/DataStruct/LockList.py
# -*- coding: utf-8 -*-

'''
@brief 线程安全列表， T 是 Object ，便于使用 Equal 比较地址
'''

from Libs.Core.GObject import GObject;
from Libs.Thread.MMutex import MMutex;
from Libs.DataStruct.MList import MList;

class LockList(GObject):

    def __init__(self):
        super(LockList, self).__init__();
        
        self.mTypeId = "LockList";
        
        self.mVisitMutex = MMutex();
        self.mList = MList();


    def Count(self): 
        self.mVisitMutex.lock();
        return self.mList.length();
        self.mVisitMutex.unlock();


    def Add(self, item):
        self.mVisitMutex.lock();
        self.mList.Add(item);
        self.mVisitMutex.unlock();

    def Remove(self, item):
        self.mVisitMutex.lock();
        self.mList.Remove(item);
        self.mVisitMutex.unlock();


    def RemoveAt(self, index):
        self.mVisitMutex.lock();
        self.mList.RemoveAt(index);
        self.mVisitMutex.unlock();


    def IndexOf(self, item):
        self.mVisitMutex.lock();
        self.mList.IndexOf(item);
        self.mVisitMutex.unlock();

File: JSLib/src/DataStruct/LockList.js
using System;

namespace SDK.Lib
{
    /**
     * @brief 线程安全列表， T 是 Object ，便于使用 Equal 比较地址
     */
    public class LockList<T>
    {
        protected DynBuffer<T> m_dynamicBuffer;
        protected MMutex m_visitMutex;
        protected T m_retItem;

        public LockList(string name, uint initCapacity = 32/*DataCV.INIT_ELEM_CAPACITY*/, uint maxCapacity = 8 * 1024 * 1024/*DataCV.MAX_CAPACITY*/)
        {
            m_dynamicBuffer = new DynBuffer<T>(initCapacity, maxCapacity);
            m_visitMutex = new MMutex(false, name);
        }

        public uint Count 
        { 
            get
            {
                using (MLock mlock = new MLock(m_visitMutex))
                {
                    return m_dynamicBuffer.m_size;
                }
            }
        }

        public T this[int index] 
        { 
            get
            {
                using (MLock mlock = new MLock(m_visitMutex))
                {
                    if (index < m_dynamicBuffer.m_size)
                    {
                        return m_dynamicBuffer.m_buff[index];
                    }
                    else
                    {
                        return default(T);
                    }
                }
            }

            set
            {
                using (MLock mlock = new MLock(m_visitMutex))
                {
                    m_dynamicBuffer.m_buff[index] = value;
                }
            }
        }

        public void Add(T item)
        {
            using (MLock mlock = new MLock(m_visitMutex))
            {
                if (m_dynamicBuffer.m_size >= m_dynamicBuffer.m_iCapacity)
                {
                    m_dynamicBuffer.extendDeltaCapicity(1);
                }

                m_dynamicBuffer.m_buff[m_dynamicBuffer.m_size] = item;
                ++m_dynamicBuffer.m_size;
            }
        }

        public bool Remove(T item)
        {
            using (MLock mlock = new MLock(m_visitMutex))
            {
                int idx = 0;
                foreach (var elem in m_dynamicBuffer.m_buff)
                {
                    if(item.Equals(elem))       // 地址比较
                    {
                        this.RemoveAt(idx);
                        return true;
                    }

                    ++idx;
                }
                return false;
            }
        }

        public T RemoveAt(int index)
        {
            using (MLock mlock = new MLock(m_visitMutex))
            {
                if (index < m_dynamicBuffer.m_size)
                {
                    m_retItem = m_dynamicBuffer.m_buff[index];

                    if (index < m_dynamicBuffer.m_size)
                    {
                        if (index != m_dynamicBuffer.m_size - 1 && 1 != m_dynamicBuffer.m_size) // 如果删除不是最后一个元素或者总共就大于一个元素
                        {
                            Array.Copy(m_dynamicBuffer.m_buff, index + 1, m_dynamicBuffer.m_buff, index, m_dynamicBuffer.m_size - 1 - index);
                        }

                        --m_dynamicBuffer.m_size;
                    }
                }
                else
                {
                    m_retItem = default(T);
                }

                return m_retItem;
            }
        }

        public int IndexOf(T item)
        {
            using (MLock mlock = new MLock(m_visitMutex))
            {
                int idx = 0;
                foreach (var elem in m_dynamicBuffer.m_buff)
                {
                    if (item.Equals(elem))       // 地址比较
                    {
                        this.RemoveAt(idx);
                        return idx;
                    }

                    ++idx;
                }
                return -1;
            }
        }
    }
}